go/ai/prompt.go

```diff
@@ -426,3 +430,149 @@ func renderDotprompt(templateText string, variables map[string]any, defaultInput
 	}
 	return str, nil
 }
+
+// LoadPromptFolder loads prompts and partials from the input directory for the given namespace
+func LoadPromptFolder(registry *registry.Registry, dir string, ns string) {
+	if dir == "" {
+		dir = "./prompts"
+	}
+	promptsPath, err := filepath.Abs(dir)
+	if err != nil {
+		log.Printf("Failed to resolve directory: %v", err)
+		return
+	}
+	if _, err := os.Stat(promptsPath); os.IsNotExist(err) {
+		log.Printf("Prompts directory does not exist: %s", promptsPath)
+		return
+	}
+	LoadPromptFolderRecursively(registry, promptsPath, ns, "")
+}
+
+// LoadPromptFolderRecursively recursively loads prompts and partials from the directory.
+func LoadPromptFolderRecursively(registry *registry.Registry, dir string, ns string, subDir string) {
+	parentPath := filepath.Join(dir, subDir)
+	entries, err := os.ReadDir(parentPath)
+	if err != nil {
+		log.Printf("Failed to read directory: %v", err)
+		return
+	}
+
+	for _, entry := range entries {
+		fileName := entry.Name()
+		if entry.IsDir() {
+			LoadPromptFolderRecursively(registry, dir, ns, filepath.Join(subDir, fileName))
+		} else if strings.HasSuffix(fileName, ".prompt") {
+			if strings.HasPrefix(fileName, "_") {
+				partialName := strings.TrimSuffix(fileName[1:], ".prompt")
+				source, err := os.ReadFile(filepath.Join(parentPath, fileName))
+				if err != nil {
+					log.Printf("Failed to read partial file: %v", err)
+					continue
+				}
+				definePartial(registry, partialName, string(source))
+				log.Printf("Registered Dotprompt partial \"%s\" from \"%s\"", partialName, filepath.Join(parentPath, fileName))
+			} else {
+				loadPrompt(registry, dir, fileName, subDir, ns)
+			}
+		}
+	}
+}
+
+// definePartial registers a partial template in the registry.
+func definePartial(registry *registry.Registry, name string, source string) {
+	// TODO: Add this functionality
+}
+
+// loadPrompt loads a single prompt into the registry.
+func loadPrompt(registry *registry.Registry, path string, filename string, prefix string, ns string) {
+	name := strings.TrimSuffix(filename, ".prompt")
+	var variant string
+	if strings.Contains(name, ".") {
+		parts := strings.SplitN(name, ".", 2)
+		name = parts[0]
+		variant = parts[1]
+	}
+
+	source, err := os.ReadFile(filepath.Join(path, prefix, filename))
+	if err != nil {
+		log.Printf("Failed to read prompt file: %v", err)
+		return
+	}
+	parsedPrompt, err := registry.Dotprompt.Parse(string(source))
+	if err != nil {
+		log.Printf("Failed to parse prompt: %v", err)
+		return
+	}
+	metadata, err := registry.Dotprompt.RenderMetadata(string(source), &parsedPrompt.PromptMetadata)
+	if err != nil {
+		log.Printf("Failed to render metadata: %v", err)
+		return
+	}
+
+	commonOpts := commonOptions{
```

Comment:
This is perfectly fine how you are doing it here but I think it would be best to use the option interface like a user, so instead of creating the options struct manually, create a list of options that you append to like:

```go
commonOpts := []CommonOption{WithModelName(metadata.Model)}

if toolChoice, ok := metadata.Raw["toolChoice"].(ToolChoice); ok {
	commonOpts = append(commonOpts, WithToolChoice(toolChoice))
}

...
```

This way if we ever add any processing in those option functions, it will take affect here too. The only one that it won't work for is the default input/schema and it's ok to just have the option like you have it there.

Reply:
Actually, strike that.

Since it's not just input, let's forego all `WithX` and create a `promptOptions` struct directly and set all options that way. I think it'll be worse having a mix long term since we'll still have to make sure that these exception fields behave the same.